zephyr/lib/regions_mm.c

```diff
@@ -1,52 +1,570 @@
+ * @retval NULL on creation failure.
  */
-static int virtual_heaps_init(void)
+struct virtual_memory_heap *vmh_init_heap(const struct heap_config *cfg,
+					  int memory_region_attribute, bool span)
 {
+	int core_id = cpu_get_id();
 	struct sys_mm_drv_region *virtual_memory_regions =
 		(struct sys_mm_drv_region *)sys_mm_drv_query_memory_regions();
+	int i;
+
+	/* Check if we haven't created heap for this region already */
+	if (vmh_get_heap_by_attribute(memory_region_attribute, core_id))
+		return NULL;
+
+	struct virtual_memory_heap *new_heap =
+	    rzalloc(SOF_MEM_ZONE_SYS_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*new_heap));
+
+	if (!new_heap)
+		return NULL;
+
+	list_init(&new_heap->node);
+	struct heap_config new_config = {0};
+
+	/* Initialize list that will keep all active allocations and
+	 * the current mappings list
+	 */
+	list_init(&new_heap->active_allocations);
+	list_init(&new_heap->active_physical_mappings_list);
+
+	/* Search for matching attribute so we place heap on right
+	 * virtual region.
+	 * In case of core heaps regions we count theme from 0 to max
+	 * available cores
+	 */
+	if (memory_region_attribute == MEM_REG_ATTR_CORE_HEAP) {
+		new_heap->virtual_region = &virtual_memory_regions[core_id];
+	} else {
+		for (i = CONFIG_MP_MAX_NUM_CPUS;
+			i < CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT; i++) {
+
+			if (memory_region_attribute == virtual_memory_regions[i].attr) {
+				new_heap->virtual_region = &virtual_memory_regions[i];
+				break;
+			}
+		}
+	}
+
+	/* If no config was specified we will use default one */
+	if (!cfg) {
+		vmh_get_default_heap_config(new_heap->virtual_region, &new_config);
+		cfg = &new_config;
+	}
+
+	/* Check if configuration provided can fit into virtual regions memory
+	 * and if it is physical page aligned
+	 */
+	size_t total_requested_size = 0;
+
+	for (i = 0; i < MAX_MEMORY_ALLOCATORS_COUNT; i++) {
+		if (cfg->block_bundles_table[i].block_size > 0) {
+			total_requested_size += cfg->block_bundles_table[i].block_size *
+						cfg->block_bundles_table[i].number_of_blocks;
+		}
+	}
+
+	if (total_requested_size > new_heap->virtual_region->size
+		|| total_requested_size % CONFIG_MM_DRV_PAGE_SIZE)
+		goto fail;
+
+	/* Offset used to calculate where in virtual region to start buffer
+	 * for the memblock
+	 */
+	uint32_t offset = 0;
 
-	for (size_t i = 0;
-		i < CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT;
-		i++) {
-		vm_heaps[i].virtual_region = &virtual_memory_regions[i];
+	/* Iterate over all blocks and fill those that have config size value
+	 * higher than 0.
+	 * Each cfg entry is converted into new allocators based on mem_blocks
+	 * API. It needs a lot of runtime allocations for members of mem_blocks.
+	 */
+	for (i = 0; i < MAX_MEMORY_ALLOCATORS_COUNT; i++) {
+		/* Only run for sizes > 0 */
+		if (!cfg->block_bundles_table[i].block_size)
+			continue;
+		/* Need to create variety of structs and constantly keep
+		 * them in memory so we need to keep them on sysheap.
+		 */
```

Comment:
this needs to be clearer than "variety of structs", i.e. what are we really doing in the next block/phase. I assume we are using the new_allocator to track some context ?